File: 75. Sort Colors.js
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var sortColors = function(nums) {
    if ( nums === null || nums.length === 0 ) {
        return null;
    }

    quickSort(nums, 0, nums.length-1);
};

var quickSort = function(nums, left, right) {
    if ( left >= right ) {
        return;
    }

    var mid = nums[Math.floor((left+right)/2)];
    var l = left;
    var r = right;

    while ( l <= r ) {
        while ( l <= r && nums[l] < mid ) {
            l++;
        }
        while ( l <= r && nums[r] > mid ) {
            r--;
        }
        if ( l <= r ) {
            var tmp = nums[l];
            nums[l] = nums[r];
            nums[r] = tmp;

            l++;
            r--;
        }
    }

    quickSort(nums, left, r);
    quickSort(nums, l, right);
}

File: 75. Sort Colors.py
class Solution(object):
    def sortColors(self, nums):
        self.quickSort(nums, 0, len(nums) - 1)

    def quickSort(self, nums, start, end):
        if start >= end:
            return

        left = start
        right = end
        pivot = nums[(start + end) / 2]

        while left <= right:
            while left <= right and nums[left] < pivot:
                left += 1
            while left <= right and nums[right] > pivot:
                right -= 1
            if left <= right:
                nums[left], nums[right] = nums[right], nums[left]
                left += 1
                right -= 1
        self.quickSort(nums, start, right)
        self.quickSort(nums, left, end)